### ISSUE TITLE
Lagging at hell !

### ISSUE THREAD

**Issue by Yuri6037** [DrChat/Gmod-vphysics#51]
I don't know how you can play GMod with this DLL because it's terribly laggy, 2~5 FPS !!!

With normal vphysics i have arround 100 FPS and with your vphysics it's unplayable (2~5 FPS).

I don't know why this is as much laggy.

Also i noticed that it's getting back to 100 FPS only on the gm_construct. But just spawn 1 AdvDupe 2 and it goes back to 2~5 FPS, and you hear a super noise in the heatset !

**Comment by DrChat** [comment]
Coolio! Any steps to reproduce? What map are you playing on? What dupes are you duping?

**Comment by Yuri6037** [comment]
Load rp_rockford_v1b and see... Second, use SBEP (Spacebuild Enhancement Project models + the GyroPod tool), and then sit down on control chair and enjoy your 3 FPS !

**Comment by DrChat** [comment]
Can you narrow this down to a more specific test case? Is it the map itself causing the lag, or is the addon doing it? Can you reproduce this issue without using addons?

**Comment by Yuri6037** [comment]
Without addons no, that's only happening when you have welded object together. For example weld about 200 objects together like you have on a Wiremod car, and then you can enjoy your 5 FPS (a bit more than with SBEP).

Anyway, your idea is realy good, but your physics engine is much less efficient than vphysics.dll...

I'd realy like to put it on my server to avoid stupid Crazy Physics crashs, but first mass values are completely broken, and you broke TDMCars, only the HL2 vehicles works fine... In all cases with the minimum addons mounted it's playable but a bit laggy...

So in conclusion your physics engine is globaly more laggy than valve's one and is buggy... You can't use it on any roleplay servers it breaks object's mass, and breaks TDMCars physics...

I'll make a video for you want to understand better the problem...

**Comment by Yuri6037** [comment]
Are you still willing to support the project ? Because I'd like to see this DLL working much better and able to be included in servers...

**Comment by DrChat** [comment]
Sorry, don't think I'll have much time for this project anymore. I'm having too much fun working on Xenia, the Xbox 360 emulator, at the moment. Anyone else can pick this project up though. There's not much left to be done, actually.

**Comment by GranPC** [comment]
RIP gmod-vphysics. F

**Comment by DrChat** [comment]
:(

**Comment by Yuri6037** [comment]
That was a good try, but well now it's done... I could take it however : 
1- I'm bad a Bullet physics
2- I need time
3- I have exams this week....
4- This will need me to enter back the C++ after programmed GLua for some monthes...

**Comment by ILOVEPIE** [comment]
I can pick this up.
Will need to learn the bullet api however.

**Comment by Gliby** [comment]
Do it!

**Comment by 60percentCaffeine** [comment]
I'm getting the same issue too. It seems that Bullet Physics (or just this implementation)  is heavier to CPU than VPhysics, even with multithreading support. That's, actually, bad. I thought this will have better perfomance than the default one. I've also noticed one more issue: on ttt_vault, under one of the stairs, there is two big boxes. They're just going crazy. If they're don't moving, just touch them - this will activate physics simulation.

**Comment by Yuri6037** [comment]
Now It's obvious that I'll no longer have any time for that. I decided to rewrite GMod from scratch, to support newer technologies (DX11 for example) using UE4.
I'm able to get much better rendering, and much better performances when it comes to physics simulation.

I could even run the game under the required specs from EpicGames.

I wrote my own way to handle Lua, giving me full control on the VM (instead of using a script plugin).

I destroyed securities to prevent low level protocol modification from C without editing source code, allowing me to write my own bytes sending channels and of course writing lua files through the UDP protocol !

**Comment by Spirrwell** [comment]
Hey, I know you're not particularly working on this project anymore. But it does have me quite interested. I'm a programmer for the free Steam Source mod Pirates, Vikings, and Knights II.

To make the long story short, I tested this out so I could potentially deal with the typical ragdoll crashes, and the major lag dips I experienced which dropped the FPS to 10 FPS stemmed from specific ragdolls. It seems the two classes in the game that are affected by this are the Heavy Knight, and Berserker.

The FPS would dip down to 10, but after a few moments when the ragdoll "settles" I guess, the FPS returns to normal. Would you happen to know what might cause that behavior? The collision model, or something else?

I'm asking because I'm wondering if it's related to some of the typical ragdoll crashes we get when using the normal vphysics implementation.

Thanks for any information you can provide!

**Comment by 60percentCaffeine** [comment]
@Yuri6037 can we look at your GMod rewrite somewhere?

**Comment by Yuri6037** [comment]
Hello,
I couldn't see this before... I did not disband the idea of a new GMod... Rewriting GMod in UE4 revealed to be MUCH more complicated than I thought first...

Someday such replacement will exist, but maybe not using UE4...
Indeed, the way Unreal works is not suitable for the type of project GMod is : I need full modding support from Lua (even the renderer) on at least Windows, Linux and Mac. As you may know already UE4 does not support Linux and Mac that well...

However, the idea is not dead ; right now I'm on a new project that will in the end support better the idea of a GMod rewrite, I can't tell anything about that right now, it's really LONG. The only thing I can tell right now is that if I can finish what I began then I can recreate a better GMod but this time with even more than only one Physics Engine : the ability to switch physics engine at runtime...
Sorry if that is not so clear, I'm working hard on it right now alongside IT school...

**Comment by GranPC** [comment]
You might want to take a look at http://www.adrianboeing.com/pal/index.html
- it's a wrapper around a bunch of different physics engines. Might help
you with the whole "switch physics engines" thing.

**Comment by Yuri6037** [comment]
Thanks but I already have my own wrapper... I just need to implement the different physics engines...
